### DISCUSSION TITLE
[mini.hipatterns] highlighting of keywords like `todo-comments`

### DISCUSSION THREAD

**Discussion by pkazmier** [nvim-mini/mini.nvim#783] [Show and tell]
Thought I'd share my highlighting rules for mini.hipatterns, which mimics folke's `todo-comments` in the event others might be interested.

Specifically, it ensures the keyword's reverse text has a single space on the left and right side of the word. The space on the right side is created by hiding the colon using the same fg and bg color. Finally, it uses the reverse color as the foreground for the remainder of the line.

```lua
      opts.highlighters = {
        hex_color   = hipatterns.gen_highlighter.hex_color(),
        fixme       = { pattern = "() FIXME():",   group = "MiniHipatternsFixme" },
        hack        = { pattern = "() HACK():",    group = "MiniHipatternsHack" },
        todo        = { pattern = "() TODO():",    group = "MiniHipatternsTodo" },
        note        = { pattern = "() NOTE():",    group = "MiniHipatternsNote" },
        fixme_colon = { pattern = " FIXME():()",   group = "MiniHipatternsFixmeColon" },
        hack_colon  = { pattern = " HACK():()",    group = "MiniHipatternsHackColon" },
        todo_colon  = { pattern = " TODO():()",    group = "MiniHipatternsTodoColon" },
        note_colon  = { pattern = " NOTE():()",    group = "MiniHipatternsNoteColon" },
        fixme_body  = { pattern = " FIXME:().*()", group = "MiniHipatternsFixmeBody" },
        hack_body   = { pattern = " HACK:().*()",  group = "MiniHipatternsHackBody" },
        todo_body   = { pattern = " TODO:().*()",  group = "MiniHipatternsTodoBody" },
        note_body   = { pattern = " NOTE:().*()",  group = "MiniHipatternsNoteBody" },
      }
```

<img width="933" alt="image" src="https://github.com/echasnovski/mini.nvim/assets/747855/b3e2db3b-ca98-4697-b893-fa3ff09c9cac">

For comparison, here is the version using `mini.extras.gen_highlighter`:

```lua
      local hi_words = require("mini.extra").gen_highlighter.words
      -- stylua: ignore
      opts.highlighters = {
        hex_color = hipatterns.gen_highlighter.hex_color(),
        fixme = hi_words({ "FIXME" }, "MiniHipatternsFixme"),
        todo  = hi_words({ "TODO" },  "MiniHipatternsTodo"),
        note  = hi_words({ "NOTE" },  "MiniHipatternsNote"),
        hack  = hi_words({ "HACK" },  "MiniHipatternsHack"),
      }
```

<img width="933" alt="image" src="https://github.com/echasnovski/mini.nvim/assets/747855/13ecd1fb-df51-46b4-a3be-b7a48289b5fc">

**Comment by echasnovski** [comment]
> Thought I'd share my highlighting rules for mini.hipatterns, which mimics folke's `todo-comments` in the event others might be interested.
> 
> Specifically, it ensures the keyword's reverse text has a single space on the left and right side of the word. The space on the right side is created by hiding the colon using the same fg and bg color. Finally, it uses the reverse color as the foreground for the remainder of the line.

Oh, that looks nice!

I am personally not a fan of highlighting whole line containing the keyword, but it certainly look nice.

**Comment by ahmedelgabri** [comment]
@pkazmier I think this can even be simplified by passing an array of patterns for the keyword + the colon 

https://github.com/echasnovski/mini.nvim/blob/a3f274633a57c713160baf51724ecfc051117877/doc/mini-hipatterns.txt#L273-L274 

Update: ignore this, it won't work in your case.

Here is what I'm doing (_I now it can be simplified, still didn't make up my mind on highlighting the whole line or not_)

```lua
local hipatterns = require 'mini.hipatterns'

local highlighters = {}
for _, word in ipairs {
	'todo',
	'note',
	'hack',
	'fixme',
	{ 'warn', 'hack' },
	{ 'bug', 'fixme' },
} do
	local w = type(word) == 'table' and word[1] or word
	local hl = type(word) == 'table' and word[2] or word

	highlighters[w] = {
		pattern = {
			string.format('()@%s()', w:upper()),
			string.format('()%s()', w:upper()),
			string.format('%s():()', w:upper()),
			string.format('%s ().*()', w:upper()),
			string.format('%s: ().*()', w:upper()),
		},
		group = string.format(
			'MiniHipatterns%s',
			hl:sub(1, 1):upper() .. hl:sub(2)
		),
	}
end

hipatterns.setup {
	highlighters = vim.tbl_extend('force', highlighters, {
		-- Highlight hex color strings (`#rrggbb`) using that color
		hex_color = hipatterns.gen_highlighter.hex_color(),
	}),
}
```

looks like this 

<img width="370" alt="Screenshot 2024-06-22 at 13 27 31" src="https://github.com/echasnovski/mini.nvim/assets/63876/c3ec57c1-f553-4768-84d8-2ef125c0b7be">
